flink-end-to-end-perf-tests/flink-basic-operations/src/main/java/org/apache/flink/basic/utils/FromElements.java

```diff
@@ -0,0 +1,235 @@
+		private Long maxCount = 0L;
+		private int sleepNum = 0;
+		private transient ListState<Long> operateState;
+		private transient Meter sourceTpsMetrics;
+
+		public FromElementsRichFunction(TypeSerializer<T> serializer, Long maxCount, int sleepNum, T... elements) throws IOException {
+			this(serializer, maxCount, sleepNum, Arrays.asList(elements));
+		}
+
+		public FromElementsRichFunction(TypeSerializer<T> serializer, Long maxCount, int sleepNum, Iterable<T> elements) throws IOException {
+			ByteArrayOutputStream baos = new ByteArrayOutputStream();
+			DataOutputViewStreamWrapper wrapper = new DataOutputViewStreamWrapper(baos);
+			int count = 0;
+			try {
+				for (T element : elements) {
+					serializer.serialize(element, wrapper);
+					count++;
+				}
+			} catch (Exception e) {
+				throw new IOException("Serializing the source elements failed: " + e.getMessage(), e);
+			}
+			this.serializer = serializer;
+			this.elementsSerialized = baos.toByteArray();
+			this.maxCount = maxCount;
+			this.sleepNum = sleepNum;
+		}
+
+		@Override
+		public void open(Configuration parameters) throws Exception {
+			RuntimeContext runtimeContext = getRuntimeContext();
+			MetricGroup metricGroup = runtimeContext.getMetricGroup();
+			final String sourceTpsName = "tps";
+			Counter sourceCounter = metricGroup.counter(sourceTpsName + "_counter", new SimpleCounter());
+			sourceTpsMetrics = metricGroup.meter(sourceTpsName, new MeterView(sourceCounter, 60));
+		}
+
+		@Override
+		public void close() {}
+
+		@Override
+		public void initializeState(FunctionInitializationContext context) throws Exception {
+			Preconditions.checkState(this.operateState == null,
+				"The " + getClass().getSimpleName() + " has already been initialized.");
+			this.operateState = context.getOperatorStateStore().getOperatorState(
+				new ListStateDescriptor<>("from-elements-state", LongSerializer.INSTANCE));
+			if (context.isRestored()) {
+				List<Long> retrievedStates = new ArrayList<>();
+				for (Long entry : this.operateState.get()) {
+					retrievedStates.add(entry);
+				}
+				this.numElementsToSkip = retrievedStates.get(0);
+			}
+		}
+
+		@Override
+		public void run(SourceContext<T> ctx) throws Exception {
+			ByteArrayInputStream bais = new ByteArrayInputStream(elementsSerialized);
+			final DataInputView input = new DataInputViewStreamWrapper(bais);
+			// if we are restored from a checkpoint and need to skip elements, skip them
+			long toSkip = numElementsToSkip;
+			if (toSkip > 0) {
+				try {
+					while (toSkip > 0) {
+						serializer.deserialize(input);
+						toSkip--;
+					}
+				} catch (Exception e) {
+					throw new IOException("Failed to deserialize an element from the source. If you are using user-defined serialization (Value and Writable types), check the serialization functions.\nSerializer is " + serializer);
+				}
+				this.numElementsEmitted = this.numElementsToSkip;
+			}
+			while (isRunning && num < maxCount) {
```

Comment:
The logic here seems to be incorrect, once `num` reached `maxCount` there won't be any element emit to the given context, is this the real purpose? Could you double check? Thanks.

Reply:
yes, the original purpose here is that when num reached maxCount, there won't be any element emit to the given context. it's no need to double check,so i removed isRunning